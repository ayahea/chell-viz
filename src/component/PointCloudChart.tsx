import * as React from 'react';

import { RESIDUE_TYPE } from '../data/chell-data';

import PlotlyChart, { defaultConfig, defaultLayout, generatePointCloudData } from '../helper/PlotlyHelper';

export interface IScatterChartProps {
  candidateResidues: RESIDUE_TYPE[];
  inputData: Array<{
    color: string;
    points: Array<{ i: number; j: number }>;
  }>;
  height: number;
  hoveredResidues: RESIDUE_TYPE[];
  nodeSize: number;
  width: number;
  onClickCallback?: () => void;
  onHoverCallback?: () => void;
  onSelectedCallback?: () => void;
  onUnHoverCallback?: () => void;
}

class ScatterChart extends React.Component<IScatterChartProps, any> {
  constructor(props: any) {
    super(props);
  }
  public render() {
    const { candidateResidues, height, hoveredResidues, inputData, nodeSize, width, ...props } = this.props;

    const pointCloudData = inputData.map(entry =>
      generatePointCloudData(this.generateFloat32ArrayFromContacts(entry.points), entry.color, nodeSize),
    );

    return (
      <PlotlyChart
        config={{
          ...defaultConfig,
        }}
        data={pointCloudData}
        layout={{
          ...defaultLayout,
          height,
          width,
          xaxis: {
            ...defaultLayout.xaxis,
            gridcolor: '#ff0000',
            gridwidth: nodeSize,
            showticklabels: false,
            tickvals: [...candidateResidues, ...hoveredResidues],
          },
          yaxis: {
            ...defaultLayout.yaxis,
            gridcolor: '#ff0000',
            gridwidth: nodeSize,
            showticklabels: false,
            tickvals: [...candidateResidues, ...hoveredResidues],
          },
        }}
        {...props}
      />
    );
  }

  protected generateFloat32ArrayFromContacts = (array: Array<{ i: number; j: number }>) => {
    const result = new Float32Array(array.length * 2);
    array.forEach((item, index) => {
      result[index * 2] = item.i;
      result[index * 2 + 1] = item.j;
    });
    return result;
  };
}

export default ScatterChart;
export { ScatterChart };
