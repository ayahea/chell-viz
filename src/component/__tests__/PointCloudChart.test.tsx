import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as React from 'react';

import PointCloudChart from '../PointCloudChart';

describe('PointCloudChart', () => {
  const emptyData = [
    {
      color: '',
      points: [],
    },
  ];

  test('Should match existing snapshot when given simple data.', () => {
    const wrapper = shallow(<PointCloudChart data={emptyData} nodeSize={4} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
