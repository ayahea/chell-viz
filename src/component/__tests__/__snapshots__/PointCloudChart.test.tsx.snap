// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PointCloudChart Should match existing snapshot when given simple data. 1`] = `
<PlotlyChart
  config={
    Object {
      "displayModeBar": true,
    }
  }
  data={
    Array [
      Object {
        "marker": Object {
          "color": "",
          "sizemax": 8,
          "sizemin": 4,
        },
        "mode": "markers",
        "type": "pointcloud",
        "xy": Float32Array [],
      },
    ]
  }
  layout={
    Object {
      "height": 400,
      "legend": Object {},
      "showlegend": false,
      "title": "",
      "width": 400,
      "xaxis": Object {
        "gridcolor": "#ff0000",
        "gridwidth": 4,
        "showticklabels": false,
        "tickvals": Array [],
      },
      "yaxis": Object {
        "autorange": "reversed",
        "gridcolor": "#ff0000",
        "gridwidth": 4,
        "showticklabels": false,
        "tickvals": Array [],
      },
    }
  }
  onClickCallback={[Function]}
  onHoverCallback={[Function]}
  onSelectedCallback={[Function]}
  onUnHoverCallback={[Function]}
/>
`;
